kjdoc/tour.md

```diff
@@ -538,13 +538,441 @@ This section describes KJ APIs that control process execution and low-level inte
+    * **Canceled producer:** If the producer is canceled, by default it will probably destroy the fulfiller without fulfilling or reject it. As described previously, the consumer will receive a non-descript exception, which is likely unhelpful for debugging. To avoid this scenario, the producer could perhaps use `.attach(kj::defer(...))` with a lambda that checks `fulfiller->isWaiting()` and rejects it if not.
+
+Because of the complexity of the above issues, it is generally recommended that you **avoid promise-fulfiller pairs** except in cases where these issues very clearly don't matter (such as unit tests).
+
+Instead, when cancellation concerns matter, consider using "adapted promises", a more-sophisticated alternative. `kj::newAdaptedPromise<T, Adapter>()` constructs an instance of the class `Adapter` (which you define) encapsulated in a returned `Promise<T>`. `Adapter`'s constructor receives a `kj::PromiseFulfiller<T>&` used to fulfill the promise. The constructor should then register the fulfiller with the desired producer. If the promise is canceled, `Adapter`'s destructor will be invoked, and should un-register the fulfiller. One common technique is for `Adapter` implementations to form a linked list with other `Adapter`s waiting for the same producer. Adapted promises make consumer cancellation much more explicit and easy to handle, at the expense of requiring more code.
+
+### Loops
+
+Promises, due to their construction, don't lend themselves easily to classic `for()`/`while()` loops. Instead, loops should be expressed recursively, as in a functional language. For example:
+
+```c++
+kj::Promise<void> boopEvery5Seconds(kj::Timer& timer) {
+  timer.afterDelay(5 * kj::SECONDS).then([&timer]() {
+    boop();
+    // Loop by recursing.
+    return boopEvery5Seconds(timer);
+  });
+}
+```
+
+KJ promises include "tail call optimization" for loops like the one above, so that the promise chain length remains finite no matter how many times the loop iterates.
+
+**WARNING!** It is very easy to accidentally break tail call optimization, creating a memory leak. Consider the following:
+
+```c++
+kj::Promise<void> boopEvery5Seconds(kj::Timer& timer) {
+  // WARNING! MEMORY LEAK!
+  timer.afterDelay(5 * kj::SECONDS).then([&timer]() {
+    boop();
+    // Loop by recursing.
+    return boopEvery5Seconds(timer);
+  }).catch_([](kj::Exception&& exception) {
+    // Oh no, an error! Log it and end the loop.
+    KJ_LOG(ERROR, exception);
+    kj::throwFatalException(kj::mv(exception));
+  });
+}
+```
+
+The problem in this example is that the recursive call is _not_ a tail call, due to the `.catch_()` appended to the end. Every time around the loop, a new `.catch_()` is added to the promise chain. If an exception were thrown, that exception would end up being logged many times -- once for each time the loop has repeated so far.
```

Comment:
I think it'd be good to drive the memory leak point home here. Would this scenario also risk a stack overflow on destruction?

Reply:
> Would this scenario also risk a stack overflow on destruction?

Yup.